diff -Nur pine4.20/imap/src/osdep/unix/Makefile pine4.20.pld/imap/src/osdep/unix/Makefile
--- pine4.20/imap/src/osdep/unix/Makefile	Sat Oct 16 11:24:00 1999
+++ pine4.20.pld/imap/src/osdep/unix/Makefile	Sat Oct 16 11:26:45 1999
@@ -363,7 +363,8 @@
 	 SPOOLDIR=/var/spool \
 	 ACTIVEFILE=/var/lib/news/active \
 	 RSHPATH=/usr/bin/rsh \
-	 BASECFLAGS="-g -O -DNFSKLUDGE"
+	 BASECFLAGS="-O2 -mpentium -DNFSKLUDGE"
+	 CC=gcc
 
 lyn:	# LynxOS
 	$(BUILD) `cat EXTRASPECIALS` OS=$@ \
diff -Nur pine4.20/pico/makefile.lnx pine4.20.pld/pico/makefile.lnx
--- pine4.20/pico/makefile.lnx	Sat Oct 16 11:29:00 1999
+++ pine4.20.pld/pico/makefile.lnx	Sat Oct 16 11:30:05 1999
@@ -38,9 +38,9 @@
 MAKE=        make
 OPTIMIZE=    # -O
 PROFILE=     # -pg
-DEBUG=       -g -DDEBUG
-
-STDCFLAGS=	-Dlnx -DJOB_CONTROL -DPOSIX -DMOUSE
+DEBUG=       # -g -DDEBUG
+CC=          gcc
+STDCFLAGS=	-Dlnx -DJOB_CONTROL -DPOSIX -DMOUSE
 CFLAGS=         $(OPTIMIZE) $(PROFILE) $(DEBUG) $(EXTRACFLAGS) $(STDCFLAGS)
 
 # switches for library building
diff -Nur pine4.20/pine/makefile.lnx pine4.20.pld/pine/makefile.lnx
--- pine4.20/pine/makefile.lnx	Sat Oct 16 11:31:38 1999
+++ pine4.20.pld/pine/makefile.lnx	Sat Oct 16 11:32:22 1999
@@ -48,7 +48,7 @@
 OPTIMIZE=    -O2
 PROFILE=     # -pg
 DEBUG=       # -g -DDEBUG
-
+CC=          gcc
 CCLIENTDIR=  ../c-client
 PICODIR=     ../pico
 
@@ -65,7 +65,7 @@
 LIBS=        $(EXTRALIBES) $(LOCLIBS) $(LDAPLIBS) $(STDLIBS) \
              `cat $(CCLIENTDIR)/LDFLAGS`
 
-STDCFLAGS=   -DLNX -DSYSTYPE=\"LNX\" -DMOUSE
+STDCFLAGS=   -DLNX -DSYSTYPE=\"LNX\" -DMOUSE
 CFLAGS=      $(OPTIMIZE) $(PROFILE) $(DEBUG) $(EXTRACFLAGS) $(LDAPCFLAGS) \
 	     $(STDCFLAGS)
 
