--- pine4.10/pine/makefile.lnx	Thu Aug 27 18:50:00 1998
+++ pine4.10.pld/pine/makefile.lnx	Sat Feb 13 12:43:20 1999
@@ -45,9 +45,9 @@
 RM=          rm -f
 LN=          ln -s
 MAKE=        make
-OPTIMIZE=    # -O2
+OPTIMIZE=    -O2
 PROFILE=     # -pg
-DEBUG=       -g -DDEBUG
+DEBUG=       # -g -DDEBUG
 
 CCLIENTDIR=  ../c-client
 PICODIR=     ../pico
@@ -60,7 +60,7 @@
 LDAPOFILES=   addrbook.o adrbkcmd.o args.o bldaddr.o init.o \
 	      other.o pine.o takeaddr.o
 
-STDLIBS=     -ltermcap
+STDLIBS=     -lncurses
 LOCLIBS=     $(PICODIR)/libpico.a $(CCLIENTDIR)/c-client.a
 LIBS=        $(EXTRALIBES) $(LOCLIBS) $(LDAPLIBS) $(STDLIBS) \
              `cat $(CCLIENTDIR)/LDFLAGS`
@@ -85,7 +85,7 @@
 pine:  $(OFILES) $(LOCLIBS)
 	echo "char datestamp[]="\"`date`\"";" > date.c
 	echo "char hoststamp[]="\"`hostname`\"";" >> date.c
-	$(CC) $(LDFLAGS) $(CFLAGS) -o pine $(OFILES) date.c $(LIBS)
+	$(CC) $(LDFLAGS) $(CFLAGS) -o pine $(OFILES) date.c $(LIBS) -lcrypt
 
 abookcpy:	abookcpy.o $(LOCLIBES)
 	$(CC) $(LDFLAGS) $(CFLAGS) -o abookcpy abookcpy.o $(LIBS)
