--- pine4.58/pine/osdep/os-lnx.h.orig	Fri Sep 12 14:28:14 2003
+++ pine4.58/pine/osdep/os-lnx.h	Fri Sep 12 14:28:36 2003
@@ -214,10 +214,6 @@
  ----*/
 #define DF_DEFAULT_PRINTER        ANSI_PRINTER
 
-/*----- Virtually all Linux distributions come with glibc 2.x. with an excellent
- * iconv implemenation */
-#define HAVE_ICONV
-
 
 /*----- The usual sendmail configuration for sending mail on Unix ------*/
 #ifndef SENDMAIL
--- pine4.58/pine/send.c.orig	Fri Sep 12 14:34:42 2003
+++ pine4.58/pine/send.c	Fri Sep 12 14:35:46 2003
@@ -8606,7 +8606,7 @@
       fs_give((void **)&pm->value);
 
 #ifndef HAVE_ICONV
-    cs = cs ? cs : ps_global->VAR_CHAR_SET);
+    cs = cs ? cs : ps_global->VAR_CHAR_SET;
 #else
     cs = cs ? cs : ps_global->VAR_SEND_CHAR_SET ?
            ps_global->VAR_SEND_CHAR_SET : ps_global->VAR_CHAR_SET;
