diff -Nru pine4.30/pine/makefile.lnx pine4.30.new/pine/makefile.lnx
--- pine4.30/pine/makefile.lnx	Sun Nov  5 20:19:09 2000
+++ pine4.30.new/pine/makefile.lnx	Sun Nov  5 20:16:06 2000
@@ -120,7 +120,7 @@
 osdep/os-lnx.c:	osdep/bld_path osdep/canacces osdep/canonicl \
 		osdep/chnge_pw osdep/coredump osdep/creatdir \
 		osdep/diskquot.non osdep/domnames osdep/err_desc \
-		osdep/expnfldr osdep/fgetpos.non osdep/filesize osdep/fltrname \
+		osdep/expnfldr osdep/fgetpos osdep/filesize osdep/fltrname \
 		osdep/fnexpand osdep/header osdep/hostname \
 		osdep/jobcntrl osdep/lstcmpnt osdep/mimedisp osdep/pipe \
 		osdep/print osdep/pw_stuff osdep/readfile osdep/debuging.tim \
diff -Nru pine4.30/pine/osdep/os-lnx.ic pine4.30.new/pine/osdep/os-lnx.ic
--- pine4.30/pine/osdep/os-lnx.ic	Sun Nov  5 20:19:09 2000
+++ pine4.30.new/pine/osdep/os-lnx.ic	Wed Jul  5 20:16:14 2000
@@ -105,7 +105,7 @@
 
 ; Most systems have the ANSI fgetpos and fsetpos functions.  For those,
 ; use fgetpos.  If a system doesn't have it use fgetpos.non.
-include(fgetpos.non)
+include(fgetpos)
 
 ; These functions are similar to popen(), but allow both an input stream 
 ; and an output buffer.
