--- ./imap/src/osdep/unix/env_unix.c.org	Sat Sep 22 01:10:39 2001
+++ ./imap/src/osdep/unix/env_unix.c	Sat Sep 22 01:11:09 2001
@@ -69,11 +69,11 @@
 static long ftp_protection = 0644;
 static long ftp_dir_protection = 0755;
 				/* default public file protection */
-static long public_protection = 0666;
-static long public_dir_protection = 0777;
+static long public_protection = 0600;
+static long public_dir_protection = 0700;
 				/* default shared file protection */
-static long shared_protection = 0660;
-static long shared_dir_protection = 0770;
+static long shared_protection = 0600;
+static long shared_dir_protection = 0700;
 static long locktimeout = 5;	/* default lock timeout */
 				/* default prototypes */
 static MAILSTREAM *createProto = NIL;
