diff -Nur pine4.20/pine/makefile.lnx pine4.20.pld/pine/makefile.lnx
--- pine4.20/pine/makefile.lnx	Sat Oct 16 11:14:52 1999
+++ pine4.20.pld/pine/makefile.lnx	Sat Oct 16 11:15:50 1999
@@ -45,9 +45,9 @@
 RM=          rm -f
 LN=          ln -s
 MAKE=        make
-OPTIMIZE=    # -O2
+OPTIMIZE=    -O2
 PROFILE=     # -pg
-DEBUG=       -g -DDEBUG
+DEBUG=       # -g -DDEBUG
 
 CCLIENTDIR=  ../c-client
 PICODIR=     ../pico
@@ -85,7 +85,7 @@
 pine:  $(OFILES) $(LOCLIBS)
 	echo "char datestamp[]="\"`date`\"";" > date.c
 	echo "char hoststamp[]="\"`hostname`\"";" >> date.c
-	$(CC) $(LDFLAGS) $(CFLAGS) -o pine $(OFILES) date.c $(LIBS)
+	$(CC) $(LDFLAGS) $(CFLAGS) -o pine $(OFILES) date.c $(LIBS) -lcrypt
 
 abookcpy:	abookcpy.o $(LOCLIBES)
 	$(CC) $(LDFLAGS) $(CFLAGS) -o abookcpy abookcpy.o $(LIBS)
