diff -Nur pine4.20/pico/osdep/os-lnx.h pine4.20.pld/pico/osdep/os-lnx.h
--- pine4.20/pico/osdep/os-lnx.h	Sat Oct 16 11:17:45 1999
+++ pine4.20.pld/pico/osdep/os-lnx.h	Sat Oct 16 11:18:51 1999
@@ -163,7 +163,7 @@
 /*
  * Place where mail gets delivered (for pico's new mail checking)
  */
-#define MAILDIR		"/usr/spool/mail"
+#define MAILDIR		"/var/mail"
 
 /*
  * What and where the tool that checks spelling is located.  If this is
diff -Nur pine4.20/pico/osdep/os-lyn.h pine4.20.pld/pico/osdep/os-lyn.h
--- pine4.20/pico/osdep/os-lyn.h	Sat Oct 16 11:19:40 1999
+++ pine4.20.pld/pico/osdep/os-lyn.h	Sat Oct 16 11:20:23 1999
@@ -109,8 +109,8 @@
 
 
 /*-------- Use terminfo database instead of termcap --------------------*/
-/* #define USE_TERMINFO */ /* use terminfo instead of termcap */
-#define USE_TERMCAP    /* use termcap */
+#define USE_TERMINFO       /* use terminfo instead of termcap */
+/* #define USE_TERMCAP */  /* use termcap */
 
 
 
@@ -154,7 +154,7 @@
 /*
  * Place where mail gets delivered (for pico's new mail checking)
  */
-#define MAILDIR		"/usr/spool/mail"
+#define MAILDIR		"/var/mail"
 
 /*
  * What and where the tool that checks spelling is located.  If this is
