--- pine4.33/imap/src/osdep/unix/env_unix.c	Tue Mar  6 16:38:48 2001
+++ pine4.33/imap/src/osdep/unix/env_unix.sz.c	Tue Mar  6 16:37:40 2001
@@ -55,16 +55,16 @@
 				/* default directory protection */
 static long dir_protection = 0700;
 				/* default lock file protection */
-static long lock_protection = 0666;
+static long lock_protection = 0600;
 				/* default ftp file protection */
 static long ftp_protection = 0644;
 static long ftp_dir_protection = 0755;
 				/* default public file protection */
-static long public_protection = 0666;
-static long public_dir_protection = 0777;
+static long public_protection = 0600;
+static long public_dir_protection = 0700;
 				/* default shared file protection */
-static long shared_protection = 0660;
-static long shared_dir_protection = 0770;
+static long shared_protection = 0600;
+static long shared_dir_protection = 0700;
 static long locktimeout = 5;	/* default lock timeout */
 				/* flock() emulator is a no-op */
 static long disableFcntlLock = NIL;
