--- ./imap/src/osdep/unix/Makefile.org	Tue Nov 26 00:50:33 2002
+++ ./imap/src/osdep/unix/Makefile	Tue Nov 26 00:52:17 2002
@@ -446,7 +446,8 @@
 	 SPOOLDIR=/var/ \
 	 ACTIVEFILE=/var/lib/news/active \
 	 RSHPATH=/usr/bin/rsh \
-	 BASECFLAGS="-g -O"
+	 BASECFLAGS="$(OPTIMIZE) -DNFSKLUDGE" \
+	 CC=gcc
 
 lyn:	# LynxOS
 	$(BUILD) `$(CAT) SPECIALS` OS=$@ \
--- ./pico/makefile.lnx.org	Tue Nov 26 00:48:46 2002
+++ ./pico/makefile.lnx	Tue Nov 26 00:49:47 2002
@@ -36,10 +36,10 @@
 RM=          rm -f
 LN=          ln -s
 MAKE=        make
-OPTIMIZE=    # -O
-PROFILE=     # -pg
-DEBUG=       -g -DDEBUG
-
+OPTIMIZE=	-O2
+PROFILE=	# -pg
+DEBUG=		# -g -DDEBUG
+CC=		gcc
 STDCFLAGS=	-Dlnx -DJOB_CONTROL -DPOSIX -DMOUSE
 CFLAGS=         $(OPTIMIZE) $(PROFILE) $(DEBUG) $(EXTRACFLAGS) $(STDCFLAGS)
 
--- ./pine/makefile.lnx.org	Tue Nov 26 00:46:28 2002
+++ ./pine/makefile.lnx	Tue Nov 26 00:48:24 2002
@@ -45,10 +45,10 @@
 RM=          rm -f
 LN=          ln -s
 MAKE=        make
-OPTIMIZE=    # -O2
+OPTIMIZE=    -O2
 PROFILE=     # -pg
-DEBUG=       -g -DDEBUG -DDEBUGJOURNAL
-
+DEBUG=       # -g -DDEBUG -DDEBUGJOURNAL
+CC=          gcc
 CCLIENTDIR=  ../c-client
 PICODIR=     ../pico
 
@@ -87,7 +87,7 @@
 pine:  $(OFILES) $(LOCLIBS)
 	echo "char datestamp[]="\"`date`\"";" > date.c
 	echo "char hoststamp[]="\"`hostname`\"";" >> date.c
-	$(CC) $(LDFLAGS) $(CFLAGS) -o pine $(OFILES) date.c $(LIBS)
+	$(CC) $(LDFLAGS) $(CFLAGS) -o pine $(OFILES) date.c $(LIBS) -lcrypt
 
 rpload:	rpload.o $(CCLIENTDIR)/c-client.a
 	$(CC) $(LDFLAGS) $(CFLAGS) -o rpload rpload.o $(LIBS)
