--- pine4.63.orig/pine/mailcmd.c.orig	2005-06-22 11:51:25.141772744 +0000
+++ pine4.63/pine/mailcmd.c	2005-06-22 11:54:14.325052976 +0000
@@ -4118,9 +4118,10 @@
 	 * the destination folder will need...
 	 */
 	context_apply(tmp, context, save_folder, sizeof(tmp));
-        save_stream = (stream && stream->dtb
-		       && stream->dtb->flags & DR_LOCAL) && !IS_REMOTE(tmp) ?
-	  stream : context_same_stream(context, save_folder, stream);
+//        save_stream = (stream && stream->dtb
+//		       && stream->dtb->flags & DR_LOCAL) && !IS_REMOTE(tmp) ?
+//	  stream : context_same_stream(context, save_folder, stream);
+	  save_stream = context_same_stream(context, save_folder, stream);
     }
 
     /* if needed, this'll get set in mm_notify */
