diff -uNr pine4.21.orig/pine/init.c pine4.21/pine/init.c
--- pine4.21.orig/pine/init.c	Wed Nov  3 18:47:26 1999
+++ pine4.21/pine/init.c	Mon Sep 25 12:06:05 2000
@@ -2042,12 +2042,12 @@
       mail_parameters(NULL, SET_USERHASNOLIFE, (void *) 1);
 
 #ifndef	_WINDOWS
-    if(F_ON(F_OLD_UNIX_DRIVER_WRITE,ps_global)){
+/*    if(F_ON(F_OLD_UNIX_DRIVER_WRITE,ps_global)){
 	extern int unix_old_algorithm;
 
 	unix_old_algorithm = TRUE;
     }
-
+*/
     if(F_ON(F_QUELL_BEZERK_TIMEZONE,ps_global))
       mail_parameters(NULL, SET_NOTIMEZONES, (void *) 1);
 #endif
diff -uNr pine4.21.orig/pine/other.c pine4.21/pine/other.c
--- pine4.21.orig/pine/other.c	Fri Nov  5 15:00:18 1999
+++ pine4.21/pine/other.c	Mon Sep 25 12:06:05 2000
@@ -10808,7 +10808,8 @@
 	 * because otherwise strip_subject wants to give us utf8 back, and
 	 * we can't handle it.
 	 */
-	if((q = mail_strip_subject(p)) != NULL){
+	q=mail_strip_subject(p);
+	if(q != NULL){
 	    t = add_roletake_escapes(q);
 	    fs_give((void **)&q);
 	}
