diff -Nru pine4.30/imap/src/osdep/unix/Makefile pine4.30.new/imap/src/osdep/unix/Makefile
--- pine4.30/imap/src/osdep/unix/Makefile	Sun Nov  5 20:05:46 2000
+++ pine4.30.new/imap/src/osdep/unix/Makefile	Sun Nov  5 20:07:16 2000
@@ -74,7 +74,7 @@
 # Commands possibly overriden by the individual port
 
 ARRC=ar rc
-CC=cc
+CC=gcc
 LN=ln -s
 RANLIB=ranlib
 
@@ -559,7 +559,7 @@
 	 SPOOLDIR=/var/spool \
 	 ACTIVEFILE=/var/lib/news/active \
 	 RSHPATH=/usr/bin/rsh \
-	 BASECFLAGS="-g -fno-omit-frame-pointer $(GCCOPTLEVEL)" \
+	 BASECFLAGS="$(OPTIMIZE) -DNFSKLUDGE" \
 	 BASELDFLAGS="-lcrypt"
 
 sl4:	# Secure Linux using libc4
