diff -uNr pine4.21.orig/imap/src/osdep/unix/unix.c pine4.21/imap/src/osdep/unix/unix.c
--- pine4.21.orig/imap/src/osdep/unix/unix.c	Thu Oct 14 20:46:33 1999
+++ pine4.21/imap/src/osdep/unix/unix.c	Thu Sep 21 14:04:17 2000
@@ -238,7 +238,7 @@
   int i,fd;
   time_t ti = time (0);
   if (!(s = dummy_file (mbx,mailbox))) {
-    sprintf (tmp,"Can't create %.80s: invalid name",mailbox);
+    snprintf (tmp,MAILTMPLEN,"Can't create %.80s: invalid name",mailbox);
     mm_log (tmp,ERROR);
   }
 				/* create underlying file */
@@ -247,7 +247,7 @@
     if ((s = strrchr (s,'/')) && !s[1]) return T;
     if ((fd = open (mbx,O_WRONLY,
 		    (int) mail_parameters (NIL,GET_MBXPROTECTION,NIL))) < 0) {
-      sprintf (tmp,"Can't reopen mailbox node %.80s: %s",mbx,strerror (errno));
+      snprintf (tmp,MAILTMPLEN,"Can't reopen mailbox node %.80s: %s",mbx,strerror (errno));
       mm_log (tmp,ERROR);
       unlink (mbx);		/* delete the file */
     }
@@ -255,18 +255,18 @@
     else if (mail_parameters (NIL,GET_USERHASNOLIFE,NIL)) ret = T;
     else {			/* initialize header */
       memset (tmp,'\0',MAILTMPLEN);
-      sprintf (tmp,"From %s %sDate: ",pseudo_from,ctime (&ti));
+      snprintf (tmp,MAILTMPLEN,"From %s %sDate: ",pseudo_from,ctime (&ti));
       rfc822_fixed_date (s = tmp + strlen (tmp));
 				/* write the pseudo-header */
-      sprintf (s += strlen (s),
+      snprintf (s += strlen (s),MAILTMPLEN,
 	       "\nFrom: %s <%s@%s>\nSubject: %s\nX-IMAP: %010lu 0000000000",
 	       pseudo_name,pseudo_from,mylocalhost (),pseudo_subject,
 	       (unsigned long) ti);
       for (i = 0; i < NUSERFLAGS; ++i) if (default_user_flag (i))
-	sprintf (s += strlen (s)," %s",default_user_flag (i));
-      sprintf (s += strlen (s),"\nStatus: RO\n\n%s\n\n",pseudo_msg);
+	snprintf (s += strlen (s),MAILTMPLEN," %s",default_user_flag (i));
+      snprintf (s += strlen (s),MAILTMPLEN,"\nStatus: RO\n\n%s\n\n",pseudo_msg);
       if ((write (fd,tmp,strlen (tmp)) < 0) || close (fd)) {
-	sprintf (tmp,"Can't initialize mailbox node %.80s: %s",mbx,
+	snprintf (tmp,MAILTMPLEN,"Can't initialize mailbox node %.80s: %s",mbx,
 		 strerror (errno));
 	mm_log (tmp,ERROR);
 	unlink (mbx);		/* delete the file */
