diff -Nur pine4.20/imap/src/osdep/unix/os_lnx.h pine4.20.pld/imap/src/osdep/unix/os_lnx.h
--- pine4.20/imap/src/osdep/unix/os_lnx.h	Sat Oct 16 11:45:28 1999
+++ pine4.20.pld/imap/src/osdep/unix/os_lnx.h	Sat Oct 16 11:46:14 1999
@@ -41,6 +41,7 @@
 #include <fcntl.h>
 #include <syslog.h>
 #include <sys/file.h>
+#include <ustat.h>
 
 
 /* Linux gets this wrong */
@@ -53,3 +54,5 @@
 #include "ftl.h"
 #include "nl.h"
 #include "tcp.h"
+
+#include "lockfix.h"
diff -Nur pine4.20/imap/src/osdep/unix/os_lnx.c pine4.20.pld/imap/src/osdep/unix/os_lnx.c
--- pine4.20/imap/src/osdep/unix/os_lnx.c	Sat Oct 16 11:45:25 1999
+++ pine4.20.pld/imap/src/osdep/unix/os_lnx.c	Sat Oct 16 11:52:06 1999
@@ -57,4 +57,6 @@
 #define fork vfork
 #include "tcp_unix.c"
 #include "gr_waitp.c"
+#undef flock
+#include "flock.c"
 #include "tz_sv4.c"
diff -Nur pine4.20/imap/src/osdep/unix/flock.c pine4.20.pld/imap/src/osdep/unix/flock.c
--- pine4.20/imap/src/osdep/unix/flock.c	Sat Oct 16 11:47:12 1999
+++ pine4.20.pld/imap/src/osdep/unix/flock.c	Sat Oct 16 11:47:54 1999
@@ -33,10 +33,16 @@
  *
  */
  
+#ifdef USE_NFS_JACKET
 #define flock fcntl_flock
 #include "flcksafe.c"		/* get safe locking routines */
 #undef flock
 #include "flockbsd.c"		/* get flock() NFS jacket */
+#else
+#include "flcksafe.c"		/* get safe locking routines */
+#include "flocksv4.c"		/* get flock() NFS jacket */
+#define fcntl_flock flock
+#endif
 
 /* Emulator for flock() call using fcntl() locking
  * Accepts: file descriptor
