diff -Nur pine4.20/pine/osdep/os-lnx.h pine4.20.pld/pine/osdep/os-lnx.h
--- pine4.20/pine/osdep/os-lnx.h	Sat Oct 16 10:47:53 1999
+++ pine4.20.pld/pine/osdep/os-lnx.h	Sat Oct 16 10:49:36 1999
@@ -60,7 +60,7 @@
    there should be a dummy disk_quota() routine in os-xxx.c so that defining
    this won't be harmful anyway.
  ----*/
-/* #define USE_QUOTAS */
+#define USE_QUOTAS
 
 
 
@@ -160,8 +160,8 @@
 
 
 /*----- System-wide config file ----------------------------------------*/
-#define SYSTEM_PINERC             "/usr/local/lib/pine.conf"
-#define SYSTEM_PINERC_FIXED       "/usr/local/lib/pine.conf.fixed"
+#define SYSTEM_PINERC             "/etc/pine.conf"
+#define SYSTEM_PINERC_FIXED       "/etc/pine.conf.fixed"
 
 
 
@@ -223,7 +223,7 @@
 
 
 /*--------- Program employed by users to change their password ---------*/
-#define	PASSWD_PROG	"/bin/passwd"
+#define	PASSWD_PROG	"/usr/bin/passwd"
 
 
 /*-------------- A couple constants used to size arrays ----------------*/
@@ -254,7 +254,7 @@
  investigated. Users tend to go in and out of Pine a few times and there
  is one file for each pine invocation
  ----*/
-#define NUMDEBUGFILES 4
+#define NUMDEBUGFILES 0
 
 /*----------------------------------------------------------------------
    The default debug level to set (approximate meanings):
