--- pine4.10/imap/src/osdep/unix/Makefile	Sun Feb 14 14:33:01 1999
+++ pine4.10.pld/imap/src/osdep/unix/Makefile	Sun Feb 14 14:38:12 1999
@@ -82,7 +82,7 @@
 # Commands possibly overriden by the individual port
 
 ARRC=ar rc
-CC=cc
+CC=gcc
 LN=ln -s
 RANLIB=ranlib
 
@@ -471,7 +471,7 @@
 	 SPOOLDIR=/var/spool \
 	 ACTIVEFILE=/var/lib/news/active \
 	 RSHPATH=/usr/bin/rsh \
-	 BASECFLAGS="-g -fno-omit-frame-pointer -O6 -DNFSKLUDGE" \
+	 BASECFLAGS="-O2 -mpentium -fomit-frame-pointer -DNFSKLUDGE" \
 	 BASELDFLAGS="-lcrypt"
 
 sl4:	# Secure Linux using libc4
