diff -Nur pine4.20/imap/src/osdep/unix/Makefile pine4.20.pld/imap/src/osdep/unix/Makefile
--- pine4.20/imap/src/osdep/unix/Makefile	Sat Oct 16 11:33:49 1999
+++ pine4.20.pld/imap/src/osdep/unix/Makefile	Sat Oct 16 11:34:53 1999
@@ -86,7 +86,7 @@
 # Commands possibly overriden by the individual port
 
 ARRC=ar rc
-CC=cc
+CC=gcc
 LN=ln -s
 RANLIB=ranlib
 
@@ -485,7 +485,7 @@
 	 SPOOLDIR=/var/spool \
 	 ACTIVEFILE=/var/lib/news/active \
 	 RSHPATH=/usr/bin/rsh \
-	 BASECFLAGS="-g -fno-omit-frame-pointer -O6 -DNFSKLUDGE" \
+	 BASECFLAGS="-O2 -mpentium -DNFSKLUDGE" \
 	 BASELDFLAGS="-lcrypt"
 
 sl4:	# Secure Linux using libc4
