diff -Nru pine4.30/imap/src/osdep/unix/Makefile pine4.30.new/imap/src/osdep/unix/Makefile
--- pine4.30/imap/src/osdep/unix/Makefile	Sun Nov  5 19:58:50 2000
+++ pine4.30.new/imap/src/osdep/unix/Makefile	Sun Nov  5 20:00:48 2000
@@ -395,7 +395,8 @@
 	 SPOOLDIR=/var/spool \
 	 ACTIVEFILE=/var/lib/news/active \
 	 RSHPATH=/usr/bin/rsh \
-	 BASECFLAGS="-g -O"
+	 BASECFLAGS="$(OPTIMIZE) -DNFSKLUDGE" \
+	 CC=gcc
 
 lyn:	# LynxOS
 	$(BUILD) `$(CAT) SPECIALS` OS=$@ \
diff -Nru pine4.30/pico/makefile.lnx pine4.30.new/pico/makefile.lnx
--- pine4.30/pico/makefile.lnx	Sun Nov  5 19:58:50 2000
+++ pine4.30.new/pico/makefile.lnx	Sun Nov  5 19:47:37 2000
@@ -38,8 +38,8 @@
 MAKE=        make
 OPTIMIZE=    # -O
 PROFILE=     # -pg
-DEBUG=       -g -DDEBUG
-
+DEBUG=       # -g -DDEBUG
+CC=          gcc
 STDCFLAGS=	-Dlnx -DJOB_CONTROL -DPOSIX -DMOUSE
 CFLAGS=         $(OPTIMIZE) $(PROFILE) $(DEBUG) $(EXTRACFLAGS) $(STDCFLAGS)
 
diff -Nru pine4.30/pine/makefile.lnx pine4.30.new/pine/makefile.lnx
--- pine4.30/pine/makefile.lnx	Sun Nov  5 19:58:50 2000
+++ pine4.30.new/pine/makefile.lnx	Sun Nov  5 19:47:37 2000
@@ -48,7 +48,7 @@
 OPTIMIZE=    -O2
 PROFILE=     # -pg
 DEBUG=       # -g -DDEBUG
-
+CC=          gcc
 CCLIENTDIR=  ../c-client
 PICODIR=     ../pico
 
